### ISSUE TITLE
Save operation chain as named operation

### ISSUE THREAD

**Issue by d47853** [gchq/gaffer-tools#288]
Once #287 is complete there should be an option to save your operation chain as a named operation.

**Comment by n288TJYRX** [comment]
I'm going to try this

**Comment by n288TJYRX** [comment]
Added this to operation-chain-component.js but not sure where to go from here.

`   vm.saveChain = function(ev) {
        events.broadcast('onPreExecute', []);
        if (!vm.canExecute()) {
            return;
        }

        if (vm.operations.length === 0) {
            error.handle("Unable to save operation chain with no operations");
            return;
        }

        var operations = [];

        addNamedOperationOperation = {
            operationName: "test named operation",
            operations: vm.operations,
            description: 'test description',
            parameters: [],
            options: [],
            score: 2
        }

        var parameter1 = {
            value: 50,
            defaultValue: 30,
            valueClass: "Integer",
            parts: {},
        }

        addNamedOperationOperation.parameters.push(parameter1)

        operations.push(addNamedOperationOperation)
             
        query.executeQuery(
            {
                class: OPERATION_CHAIN_CLASS,
                operations: operations
            },
            function(data) {
                submitResults(data);
            }
        );

        var confirm = $mdDialog.confirm()
            .title('Operation chain saved as named operation!')
            .textContent('You can now see your saved operation in the list of operations')
            .targetEvent(ev)
            .ok('Ok')
            .cancel('Cancel');

        $mdDialog.show(confirm).then(function() {
        }, function() {
            // do nothing if they don't want to reset
        });

    }`

**Comment by n288TJYRX** [comment]
Hi, while working on this I've noticed that Delete Named Operation in the UI doesn't seem to work.

**Comment by n288TJYRX** [comment]
Okay I've now got a save button that saves the operation chain the user enters as "test named operation2" and can then load it up and execute it later. I'm now going to add an input on a pop up dialog when the user clicks on save so that they can enter a name for their operation chain. But I was wondering about other parameters and options for the named operation. How do you want me to implement this? Or for now should I just leave the named operations without any parameters and options.

**Comment by d47853** [comment]
Yeah parameters can be a later iteration

**Comment by p013570** [comment]
@d47853, @n288TJYRX - looks like this code has been merged into 1.10-RC1 - should this ticket be closed?
The feature looks really useful! It works really well - thanks!

However, I am concerned that users may not be aware that it saved query will be globally available to all other users. It also doesn't set the write access roles on the query, so someone else could potentially change their saved query without them realising.

You may already have a plan for this that I am not aware of but my recommendation would be:

- To set the read and write roles to the current user. So for the time being only the current user could access it and run it. 
- In the future you could then add a button to expose it to other users.
- I would added the user's name (or 'short' name) as a prefix to the operation name - otherwise things will very quickly start clashing.
- Perhaps add to extra documentation in the UI to explain what "saving" an operation chain actually does - i.e who will have access to it, will it save the user's seeds, will it save the user's filters, how they might go about editing/deleting it.
- I don't think there is currently a nice way to check if a user has permission to run AddNamedOperation. If they don't then why then try to save the operation they do get a nice useful message: "User does not have permission to run operation: uk.gov.gchq.gaffer.named.operation.AddNamedOperation" however in the future it would be nice to just hide the "Save Operation Chain" button.

**Comment by d47853** [comment]
Thanks @p013570 for your comments. We really just did this as a first iteration, it's great to get feedback so early.

* For name clashes, the UI isn't currently aware of who the user is, so adding any kind of username specific features would require changes to the REST API.
* With regard to publishing this globally, I was under the impression that not adding read/write access roles made the named operation private - but Gaffer's standard user factory just hardcodes the user Id.
* I think you're right about documenting this feature better. This should be done before it's put in a proper release.
* I agree that it would be good to hide the "save operation chain" button / component if the user can't access it. We may need to add a way to check whether users have access to operations. For now we could add a simple configuration option to disable it. For systems that tightly control named operations, this may be better.

**Comment by d47853** [comment]
Merged into develop
